src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp

```diff
@@ -0,0 +1,151 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file FlightTaskManualAcceleration.cpp
+ */
+
+#include "FlightTaskManualAcceleration.hpp"
+#include <mathlib/mathlib.h>
+#include <float.h>
+
+using namespace matrix;
+
+bool FlightTaskManualAcceleration::activate(vehicle_local_position_setpoint_s last_setpoint)
+{
+	bool ret = FlightTaskManual::activate(last_setpoint);
+	_velocity_setpoint.zero();
+	return ret;
+}
+
+bool FlightTaskManualAcceleration::update()
+{
+	// maximum commanded acceleration and velocity
+	Vector3f acceleration_scale;
+	Vector3f velocity_scale;
+	acceleration_scale(0) = acceleration_scale(1) = _param_mpc_acc_hor.get();
+	velocity_scale(0) = velocity_scale(1) = _param_mpc_vel_manual.get();
+
+	if (_velocity(2) < 0.f) {
+		acceleration_scale(2) = _param_mpc_acc_up_max.get();
+		velocity_scale(2) = _param_mpc_z_vel_max_up.get();
+
+	} else {
+		acceleration_scale(2) = _param_mpc_acc_down_max.get();
+		velocity_scale(2) = _param_mpc_z_vel_max_dn.get();
+	}
+
+	acceleration_scale *= 2.f; // because of drag the average aceleration is half
+
+	// Yaw
+	_position_lock.updateYawFromStick(_yawspeed_setpoint, _yaw_setpoint,
+					  _sticks_expo(3) * math::radians(_param_mpc_man_y_max.get()), _yaw, _deltatime);
+
+	// Map stick input to acceleration
+	Vector2f stick_xy(&_sticks_expo(0));
+	_position_lock.limitStickUnitLengthXY(stick_xy);
+	_position_lock.rotateIntoHeadingFrameXY(stick_xy, _yaw, _yaw_setpoint);
+
+	_acceleration_setpoint = Vector3f(stick_xy(0), stick_xy(1), _sticks_expo(2)).emult(acceleration_scale);
+
+	// Add drag to limit speed and brake again
+	Vector3f cv = acceleration_scale.edivide(velocity_scale);
+	_acceleration_setpoint -= cv.emult(_velocity);
+
+	lockAltitude();
+	lockPosition(stick_xy.length());
```

Comment:
`lockAltitude` knows that it has to take `fabsf(_stick_expo(2))` but you pass `stick_xy.length()` to `lockPosition`. Since those functions are quite similar, they should also get similar parameters.